commit 373b0c1662f735b36d37360ac3860e90cc44191b (refs/changes/73/4773/2)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-08-31T11:42:30-07:00 (1 year, 1 month ago)
    
    TRITON-745 CNAPI should be able to post sysinfo after updating NICs instead of requiring Ur

diff --git a/lib/cnapi.js b/lib/cnapi.js
index fb0691f..0096d28 100644
--- a/lib/cnapi.js
+++ b/lib/cnapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -185,6 +185,17 @@ function pollTasks(job, callback) {
                 }
 
                 if (task.status == 'complete') {
+                    if (task.history &&
+                        task.history[0] &&
+                        task.history[0].event) {
+
+                        job.returned_sysinfo = task.history[0].event.sysinfo;
+                    }
+
+                    job.log.debug({
+                        sysinfo: job.returned_sysinfo
+                    }, 'task complete');
+
                     job.params.taskSuccesses.push(detail);
                     return cb(null);
                 }
@@ -216,10 +227,22 @@ function pollTasks(job, callback) {
  * - job.params.taskFailures {Array} : task ID objects of failures
  */
 function refreshServerSysinfo(job, callback) {
+    var body = {};
     var cnapi = restify.createJsonClient({ url: cnapiUrl});
     var serverUrl = '/servers/' + job.params.server_uuid + '/sysinfo-refresh';
 
-    cnapi.post(serverUrl, {}, function (error, req, res) {
+    if (job.returned_sysinfo) {
+        // If the job already got the updated sysinfo as a result of it being
+        // returned by the task we just ran, use that so we don't bother with
+        // another round of wf -> CNAPI -> Ur -> ur-agent -> sysinfo and back.
+
+        serverUrl = '/servers/' + job.params.server_uuid + '/sysinfo';
+        body = {sysinfo: job.returned_sysinfo};
+        job.log.debug({sysinfo: job.returned_sysinfo},
+            'POSTing sysinfo received from task');
+    }
+
+    cnapi.post(serverUrl, body, function _sysinfoPosted(error, req, res) {
         if (error) {
             job.log.info('Error refreshing server sysinfo');
             job.log.info(error.stack.toString());
diff --git a/package.json b/package.json
index 6dec81a..2b366cd 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "wf-shared",
   "description": "Workflow API shared tasks",
-  "version": "1.0.0",
+  "version": "1.1.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "main": "./lib/index.js",
